python/ray/_private/runtime_env/__init__.py

```diff
@@ -1,3 +1,48 @@
+import json
+import os
+import sys
+from typing import Any, Dict
+
+
+def runtime_env_override_hook(runtime_env: Dict[str, Any]) -> Dict[str, Any]:
+    """Hook to override the runtime environment from RAY_OVERRIDE_RUNTIME_ENV_JSON."""
+
+    if "RAY_OVERRIDE_RUNTIME_ENV_JSON" in os.environ:
+        runtime_env = json.loads(os.environ["RAY_OVERRIDE_RUNTIME_ENV_JSON"])
+
+    return runtime_env
+
+
+def uv_run_runtime_env_hook(runtime_env: Dict[str, Any]) -> Dict[str, Any]:
+    """Hook that detects if the driver is run in 'uv run' and sets the runtime environment accordingly."""
+
+    import argparse
+    import psutil
+
+    parent = psutil.Process().parent()
+    cmdline = parent.cmdline()
+    if cmdline[0] != "uv" or cmdline[1] != "run":
+        # This means the driver was not run with 'uv run' -- in this case
+        # we leave the runtime environment unchanged
+        return runtime_env
+
+    # Parse known arguments of uv run that impact the runtime environment
+    uv_run_parser = argparse.ArgumentParser()
+    uv_run_parser.add_argument("--directory", nargs="?")
+    known_args, unknown_args = uv_run_parser.parse_known_args(cmdline)
+
+    # Extract the arguments of 'uv run' that are not arguments of the script
+    uv_run_args = cmdline[: len(cmdline) - len(sys.argv)]
+    runtime_env["py_executable"] = " ".join(uv_run_args)
```

Comment:
how are you differntiating between uv run args vs. script args? Is this something uv is handling under the hood by not modifying sys.argv?

Reply:
the sys.argv are only the script args (they get passed to the script, which is a subprocess of uv run, by uv run), and cmdline are the full args with both the uv run args and the script args. So `uv run` args are computed in the line `cmdline[: len(cmdline) - len(sys.argv)]` above by removing the script args (in sys.argv) from all the args (cmdline)